### ISSUE TITLE
Feature: combine `:buffer` with `:open -t`

### ISSUE THREAD

**Issue by lucc** [qutebrowser/qutebrowser#3194]
It would be great if we had a `:buffer` command which could combine its functionality with `:open -t` (or `:open -w` for that matter). Weather this should be a new command or a option to one of the existing commands, I don't know.

It would provide the following features:
* completion of open buffers, bookmarks and history items (the union of the completion of `:open` and `:buffer`)
* if a buffer is selected behave like `:buffer`
* if no buffer is selected behave like `:open -t` or `:open -w` (maybe a new command would need an option for this)
* if the match is ambiguous *I* think it should behave like `:buffer` (only if no buffer can be matched any longer it should behave like `:open`)

**Comment by Enzime** [comment]
If we implement #836, this would allow us to add open buffers as a configurable category to `:open`.

**Comment by lucc** [comment]
That is only half of this idea. The open command completes urls and the action it takes is to open them.

My proposal does not specify if the urls or buffer identifiers should be completed for open buffers. *But* the action should be different from :open if the argument is a buffer identifier or an url that is already open in a buffer. That is not covered by #836 as I read it because :open would just open the url a second time.

**Comment by Enzime** [comment]
Yeah, I definitely think if we were to add open buffers to `:open` it should switch to that tab, which would be a separate action to other categories.

**Comment by The-Compiler** [comment]
Note that commands don't know about the completion (and I think they shouldn't), so `:open foo` would need to decide whether to select a buffer (or quickmark, or do a search) only based on the text `foo`.

**Comment by rcorre** [comment]
On #836 I mentioned that a separate command would help disambiguate, but I realized there's really no need to do so. `open 1/8` navigates to http://0.0.0.1/8. I think it would be fine to interpret `[0-9]+/[0-9]+` as a win/tab id.

**Comment by The-Compiler** [comment]
Yeah, that'd be fine for window/tab indexes, but I guess the main usecase is that you can do `:open foo` like you can do `:buffer foo` to navigate to an open tab with `foo` in the title.

**Comment by lucc** [comment]
From the discussion I take it you generally like the idea.

**decision one** (syntax of the command/option (strictly speaking this list already contains some semantics for the options)):
* new command (what will it be called?), it should support `-t` and `-w` like `:open`, maybe even `-p` and `-b`
* as an argument to `:open`: `-b` is taken so maybe `-B` for "buffer" or `-j` for "jump".
* as an argument to `:buffer` (which itself might be renamed soon). Maybe `-o`, `-t` and `-w` to have similar options as with plain `:open` and `-t`, `-w`.

**decision two** (semantics of the argument):
Some cases seem clear to me, some seem more complicated (opinionated):
* anything resembling a window-buffer number (matching the regex `^((\d+)/)?(\d+)$`) should be treated as a window-buffer number, potentially resulting in an error (it is unlikely that someone wants to `:open 2/3` and it would fail in most cases anyways)
* anything resembling a full url should first be matched against open buffers:
    * if an exact match is found we should jump to it
    * *for non exact matches I am not sure* (both interpretations make sense to me)
    * if no match can be found it should be `:open`ed
* anything not resembling a full url
    (* if it contains a space `:buffer` would not accept it anyways so we can just `:open` it, but the next two points subsume this one)
    *  if a matching buffer can be found -> jump to it
    * otherwise treat it like with `:open` potentially expanding search engines etc.

Most of these semantics for the arguments can be achieved if we try to call `:buffer` on the argument first and if that throws an error `:open` it.

**Comment by mschilli87** [comment]
> * anything resembling a full url should first be matched against open buffers:
>   * [...]
>   * for non exact matches I am not sure (both interpretations make sense to me)*

In this case, why not pick one as a default (doesn't really matter which one) but allow both, configurable by a new option?

**Comment by lucc** [comment]
Here is a simple proof of concept on to of v1.1.0-169-g6f028e9ad (current master)
```diff
diff --git a/qutebrowser/browser/commands.py b/qutebrowser/browser/commands.py
index 0ebcf3240..051b477e4 100644
--- a/qutebrowser/browser/commands.py
+++ b/qutebrowser/browser/commands.py
@@ -1077,6 +1077,16 @@ class CommandDispatcher:
 
         return (tabbed_browser, tabbed_browser.widget(idx-1))
 
+
+    @cmdutils.register(instance='command-dispatcher', scope='window',
+                       maxsplit=0)
+    @cmdutils.argument('text')
+    def switch_buffer_or_open(self, text=None):
+        try:
+            self.buffer(index=text)
+        except cmdexc.CommandError:
+            self.openurl(url=text, tab=True)
+
     @cmdutils.register(instance='command-dispatcher', scope='window',
                        maxsplit=0)
     @cmdutils.argument('index', completion=miscmodels.buffer)
```
Missing:
* completion
* the subtile precedence rules discussed above

**Comment by lucc** [comment]
I expanded the above into its own branch: https://github.com/lucc/qutebrowser/tree/feature/open-buffer-command

It is still only a proof of concept but I am using it successfully! I have done a `:bind o set-cmd-text -s :switch-buffer-or-open` and `:bind O set-cmd-text -s :switch-buffer-or-open-tab`.

So as soon as some of central questions are answered (how should the command/argument be named, where should the code go) I can open a pull request for easier review and testing.  Or should I open the PR with my POC code?

**Comment by rcorre** [comment]
I think [decision one](https://github.com/qutebrowser/qutebrowser/issues/3194#issuecomment-339677254) is the tough one. @The-Compiler [expressed a preference](https://github.com/qutebrowser/qutebrowser/issues/836#issuecomment-339635097) for flags over commands. However, there's a fair bit of work in the way of filtering completions based on flags, namely [replacing the command parser](https://github.com/qutebrowser/qutebrowser/issues/836#issuecomment-342215317) and [changing the model update logic](https://github.com/qutebrowser/qutebrowser/issues/836#issuecomment-342262905).

**Comment by lucc** [comment]
After testing my POC some more I must say I also tend to prefer flags. The reason is that I currently have the above mentioned bindings for `o` and `O` and so it is a lot more typing to change a command line if I notice that it did/would send me to an open buffer if I instead want to search for a wiki page for example.

**Comment by lucc** [comment]
I have rebased my branch https://github.com/lucc/qutebrowser/tree/feature/open-buffer-command with the POC code on top of v1.5. I am using it with 
```
:bind o set-cmd-text -s :switch-buffer-or-open
:bind O set-cmd-text -s :switch-buffer-or-open-tab
```

**Question:** Is it OK if I open a PR before we have decided all the points that came up above? I would make a best effort decision which can then be further discussed in the PR.

**Comment by The-Compiler** [comment]
Feel free to open a PR right now if you already have the code. However, I'm already drowning in open PRs, and I really need to get started on [some bigger refactorings](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2018-September/000051.html) soon - so I can't promise I'll be able to take a look unfortunately, and it's likely to conflict again at some point.

**Comment by lucc** [comment]
@The-Compiler  do you prefer #4955 or this approach? I want to try and help move this forward but I see no value in following two paths.

@jcmuller are you still working on #4955?

**Comment by emad-elsaid** [comment]
I have seen multiple issues and PRs for this feature. and the discussion seems to have stopped a couple of years ago. should we revive this topic or did someone find a way to do this another way?
